layers: Add a MoveToLayer(n) key

algernon opened the pull request:
While `ShiftToLayer` and `LockLayer` activate the given layer, and keep all others active as well, `MoveToLayer` activates the given layer, and deactivates all others. This allows one to have discrete layers that stand on their own, without any other layers interfering.

Basically, this is a different way to work with layers, a less powerful, but also simpler one.

Fixes #564.

obra commented:
The theory is fine. But this definitely needs documentation before I’d be comfortable merging it. if we don’t already have suitable Shift and Lok documentation then we will need to write that too.

> On Dec 31, 2019, at 7:28 AM, Gergely Nagy <notifications@github.com> wrote:
> 
> ﻿
> While ShiftToLayer and LockLayer activate the given layer, and keep all others active as well, MoveToLayer activates the given layer, and deactivates all others. This allows one to have discrete layers that stand on their own, without any other layers interfering.
> 
> Basically, this is a different way to work with layers, a less powerful, but also simpler one.
> 
> Fixes #564.
> 
> You can view, comment on, or merge this pull request online at:
> 
>   https://github.com/keyboardio/Kaleidoscope/pull/779
> 
> Commit Summary
> 
> layers: Add a MoveToLayer(n) key
> File Changes
> 
> M src/kaleidoscope/key_defs_keymaps.h (8)
> M src/kaleidoscope/layers.cpp (4)
> Patch Links:
> 
> https://github.com/keyboardio/Kaleidoscope/pull/779.patch
> https://github.com/keyboardio/Kaleidoscope/pull/779.diff
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub, or unsubscribe.

algernon commented:
I don't think we do, will write docs for them all. I've already written a bit on layers for Luis, so will just reword that and put into `docs/layers.md`.

algernon commented:
Hm... Having talked more about this with Luis, there are a number of issues with this approach. I'm closing this PR for now, and will explain my thoughts in #564 instead.

algernon commented:
After experimenting with another approach, I came to the conclusion that _this_ is the right one. Reopening, will add documentation soon.

algernon commented:
Wrote a bit of documentation about layers, layer keys, transparency, and a few other related things. I tried to keep it short and simple, not sure how well I succeeded.

obra commented on src/kaleidoscope/key_defs_keymaps.h:
```diff
@@ -21,6 +21,7 @@
 #include "kaleidoscope_internal/deprecations.h"
 
 static const uint8_t LAYER_SHIFT_OFFSET = 42;
+static const uint8_t LAYER_MOVE_OFFSET = LAYER_SHIFT_OFFSET + 33;
```
What's with the 33?

obra replied:
(I presume "we are assuming no more than 32 layers")
But in that case, shift and move offsets should be the same number. and the 42 should probably get lifted to a named constant taht explains itself.

algernon replied:
Yes, it's no more than 32 layers. I'll use 42 as the offset for move as well then. (If I could go back in time, I'd use 33, but I can only warp through the screen, not time :/)

As for the name, how about `LAYER_OP_OFFSET`? Or just `LAYER_OFFSET`...

obra replied:
LAYER_OP_OFFSET feels less likely to step on something else outside of the scope of this file

obra commented on doc/layers.md:
```diff
@@ -0,0 +1,128 @@
+# Layers
```
I think this doc could benefit from a quick start guide with "If you want to do "X" use Lock, if you want to do "Y", use Shift, if you want to do "Z", use "Move"

There's some other prose stuff, but I'm happy to tackle that by editing the document.

algernon replied:
There's a section that is almost your quick start guide... I can move that further up and tweak it a bit to be more like a quick start.

algernon commented:
Rebased on top of master, moved the documentation to `docs/`, and added a "How do I...?" quickstart section. Also introduced `LAYER_OP_OFFSET`, and made both `LAYER_SHIFT_OFFSET` and `LAYER_MOVE_OFFSET` use it.

algernon commented:
There are a few interaction issues I'm ironing out still. Will report back when I'm done.

What I'm seeing currently is that when I use the `Colormap` plugin too, moving 0->1->0, they keymap correctly moves, but the colormap plugin will show layer1 colors. This happens because we use `activate()` to activate the layer, after clearing the state. This, however, does not update the cached top layer index! We need to update that, and possibly other caches too.

algernon commented:
Updated the branch with the fix: `Layer.move()` was relying on `activate()` to update the caches, but `activate()` was assuming we also call `deactivate()` to turn layers off, which we did not do in `Layer.move()`, where we updated the layer state directly. Since we update the layer state directly, we should not rely on `activate()` either, and do the update ourselves - and unlike `activate()`, do it unconditionally, because we want to update the caches in every case.

This fix is independent of `MoveToLayer()`, and fixes all uses of `Layer.move()`. I can pull it out into a separate PR if need be.

obra commented:
Thinking about naming for this one a bit.

moveToLayer()
switchToLayer()
goToLayer()

feelings?

obra commented:
Thinking about naming for this one a bit.

moveToLayer()
switchToLayer()
goToLayer()

feelings?

algernon commented:
I'm not a fan of `goToLayer()`, both `switchTo` and `moveTo` feel.... stronger? Of those two, I'd prefer `moveToLayer`, because we already have `Layer.move()`, which the new keycode makes use of. But I'm also fine with `switchToLayer()` - but in that case, we should probably rename `Layer.move()` too.

obra commented:
Thinking about it, ShiftToLayer and SwitchToLayer read very similarly. I'm
ok sticking with moveToLayer.

Is this PR good to go?
ᐧ

On Wed, Feb 5, 2020 at 6:07 PM Gergely Nagy <notifications@github.com>
wrote:

> I'm not a fan of goToLayer(), both switchTo and moveTo feel.... stronger?
> Of those two, I'd prefer moveToLayer, because we already have Layer.move(),
> which the new keycode makes use of. But I'm also fine with switchToLayer()
> - but in that case, we should probably rename Layer.move() too.
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/keyboardio/Kaleidoscope/pull/779?email_source=notifications&email_token=AAALC2CU4L3NVK5YCQ24HZ3RBNWFTA5CNFSM4KBVGF6KYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEK5VHRY#issuecomment-582702023>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAALC2G77MM55ALMZVCXG63RBNWFTANCNFSM4KBVGF6A>
> .
>

algernon commented:
Yup, it should be good to go.
